M2/Macaulay2/packages/HomologicalShiftIdeals.m2

```diff
@@ -0,0 +1,1001 @@
+      isFullySupported(I)
+    Example
+      S = QQ[x_1..x_4];
+      I = ideal(x_1^2,x_3,x_2*x_4);
+      isFullySupported(I)
+///);
+
+isFullySupported = method(TypicalValue=>Boolean)
+isFullySupported (Ideal) := I -> (
+        if not isMonomialIdeal(I) then error
+        else n:=numgens ring I;
+        suppI:=supportIdeal I;
+        if #suppI==n then return true
+        else return false
+    );
+
+-------------------------------------------------------------------------------------------
+--  Compute the monomial with a given multidegree
+-------------------------------------------------------------------------------------------
+mydoc = concatenate(mydoc,///
+Node
+  Key
+    toMonomial
+    (toMonomial,Ring,List)
+  Headline
+    compute the monomial with a given multidegree
+  Usage
+    toMonomial(S,l)
+  Inputs
+    S:Ring
+      a polynomial ring $S = K[x_1,\ldots,x_n]$, $K$ a field
+    l:List
+      the multidegree $(a_1,\dots,a_n)$
+  Outputs
+    u:RingElement
+      the monomial $x_1^{a_1}\cdots x_n^{a_n}\in S$
+  Description
+    Text
+      The multidegree of a monomial $u\in S=K[x_1,\ldots,x_n]$ is the vector $(a_1,a_2,\dots,a_n)$ where $a_i=\max\{j:x_i^j\ \text{divides}\ u\}$.
+    Example
+      S = QQ[x_1..x_4];
+      l = {0,2,1,2};
+      toMonomial(S,l)
+///);
+
+toMonomial = method(TypicalValue=>RingElement)
+toMonomial (Ring,List) := (S,l) -> (
+	n:=numgens S-1;
+	u:=1;
+        j:=0;
+        if not(#l==n+1)==true then return "expected a multidegree whose size equals the number of variables of the polynomial ring"
+        else for i from 0 to n do (
+            j=l#i;
+            u=u*S_i^j;
+        );
+        u
+    );
+
+-------------------------------------------------------------------------------------------
+--  Compute the multidegree of a given monomial
+-------------------------------------------------------------------------------------------
+mydoc = concatenate(mydoc,///
+Node
+  Key
+    toMultidegree
+    (toMultidegree,RingElement)
+  Headline
+    compute the multidegree of a monomial
+  Usage
+    toMultidegree u
+  Inputs
+    u:RingElement
+      a monomial $u=x_1^{a_1}\cdots x_n^{a_n}$
+  Outputs
+    l:List
+      the multidegree $(a_1,\dots,a_n)$ of $u$
+  Description
+    Text
+      The multidegree of a monomial $u\in S=K[x_1,\ldots,x_n]$ is the vector $(a_1,a_2,\dots,a_n)$ where $a_i=\max\{j:x_i^j\ \text{divides}\ u\}$.
+    Example
+      S = QQ[x_1..x_4];
+      u = x_2^2*x_3*x_4^2;
+      toMultidegree(u)
+///);
+
+toMultidegree = method(TypicalValue=>List)
+toMultidegree (RingElement) := u -> (
+	S:=ring u;
+        if not isMonomialIdeal(ideal(u)) then error
+        else n:=numgens S-1;
+        L:={};
+        for i from 0 to n do (
+              mdeg:={};
+              for j from 0 to n do (
+                     if j==i then mdeg=append(mdeg,1)
+                     else mdeg=append(mdeg,0);
+              );
+              L=append(L,mdeg);
+        );
```

Comment:
Repeatedly appending to a list inside a for loop isn't very efficient.  That's what `for ... list ...` statements are for.  For example, this could be simplified to:

```m2
L := for i to n list for j to n list if i == j then 1 else 0
```

Reply:
I replaced with the (I believe) simpler command:
L:=entries id_(S^n);

Reply:
I go back to your suggestion

Reply:
`entries id_(S^n)` probably didn't work since the `Degrees` option expects a list of lists of integers, but it gave you a list of lists of elements of the ring S.  `entries id_(ZZ^n)` would work, though.  But the for loop is quicker:

```m2
i1 : elapsedTime for i to 100 list for j to 100 list if i == j then 1 else 0;
 -- .00483555s elapsed

i2 : elapsedTime entries (id_(ZZ^100));
 -- .0319489s elapsed
```